### PR TITLE
Df fix read element cache

### PR DESCRIPTION
**Pull request by jbrezmorf**

### CODE REVIEW

File: tests/03_generic_output/45_P0_data_DG_read.yaml
Comment (jbrezmorf) [inline tests/03_generic_output/45_P0_data_DG_read.yaml:1]: Why we need to set these default values?

**Review by jbrezmorf — COMMENTED** [review]
Related to the issue #1499.
 
- see which tests could be affected by bug in FieldFE boundary reader. in `02*/`
- try to reproduce the problem by an integration test

**Comment by jbrezmorf** [comment]
TODO in the `field_fe_test.cpp` was part of the pull request [#1493](https://github.com/flow123d/flow123d/pull/1493)

File: src/fields/field_algo_base.hh [hunk 1]
```diff
@@ -100,8 +100,9 @@ struct FieldAlgoBaseInitData {
 
 /// Declaration of exception.
 TYPEDEF_ERR_INFO(EI_Field, std::string);
+TYPEDEF_ERR_INFO(EI_File, std::string);
 DECLARE_INPUT_EXCEPTION(ExcUndefElementValue,
-        << "Values of some elements of FieldFE " << EI_Field::qval << " is undefined.\n"
+        << "Values of some elements of FieldFE " << EI_Field::qval << " in file " << EI_File::qval << " are undefined.\n"
```
Comment (jbrezmorf) [inline src/fields/field_algo_base.hh:105]: FieldFE <field> on region <region> have invalid value <value>.\n Provided by file <file> at element ID <elid>.

File: src/fields/field_fe.hh [hunk 2]
```diff
@@ -195,23 +195,55 @@ private:
...
+        }
+
+        /// Sets invalid state, store invalid element id and value
+        void set(unsigned int elm_id, double value) {
+            is_nan_ = true;
+            elm_id_ = elm_id;
+            value_ = value;
+        }
+
+        bool is_nan_;
```
Comment (jbrezmorf) [inline src/fields/field_fe.hh:225]: rename to is_invalid

File: src/fields/field_fe.hh [hunk 3]
```diff
@@ -195,23 +195,55 @@ private:
...
+	 * and other information (element id an value) are stored. Check of invalid values is performed
+	 * during processing data of reader cache and possible exception is thrown only if FieldFE is defined
+	 * on appropriate region.
+	 *
+	 * Data of all regions are stored in vector of RegionValueErr instances.
+	 */
+    class RegionValueErr {
+    public:
+        /// Constructor
+        RegionValueErr() : is_nan_(false) {}
```
Comment (jbrezmorf) [inline src/fields/field_fe.hh:211]: Just use two constructors: default constructor for the valid and similar to `set` method for invalid.

File: src/fields/field_fe.hh [hunk 4]
```diff
@@ -295,6 +329,15 @@ private:
     std::array<FEItem, 4> fe_item_;
     MixedPtr<FiniteElement> fe_;
 
+    /// Pointer to computational mesh
+    const Mesh *comp_mesh_;
```
Comment (jbrezmorf) [inline src/fields/field_fe.hh:333]: Isn't computational mesh part of parent FieldAlgo Base or part of DOF handler?

File: src/io/element_data_cache.cc [hunk 5]
```diff
@@ -255,38 +257,38 @@ void ElementDataCache<T>::normalize(unsigned int idx, unsigned int divisor) {
...
-    ASSERT(check_scale_data_ == CheckScaleData::check).warning("Data should be checked before scaling. Rather call 'check_values'!\n");
-
-    std::vector<T> &vec = *( this->data_.get() );
-    for(unsigned int i=0; i<vec.size(); ++i) {
-        vec[i] *= coef;
-    }
-
-    check_scale_data_ = CheckScaleData::scale;
-}
+//template <typename T>
```
Comment (jbrezmorf) [inline src/io/element_data_cache.cc:260]: delete

File: src/io/element_data_cache.hh [hunk 6]
```diff
@@ -146,26 +146,26 @@ public:
...
-    CheckResult check_values(double default_val, double lower_bound, double upper_bound);
-
-    /**
-     * Scale data vector of given 'component_idx' with scale 'coef'.
-     *
-     * Method is executed only once and should be called after check_values method.
-     * Method can be used e. g. for convert between units.
-     */
-    void scale_data(double coef);
+//    /**
```
Comment (jbrezmorf) [inline src/io/element_data_cache.hh:149]: delete

File: src/io/element_data_cache_base.hh [hunk 7]
```diff
@@ -277,6 +282,9 @@ protected:
 
     /// Is true for DummyElementDataCache
     bool is_dummy_;
+
```
Comment (jbrezmorf) [inline src/io/element_data_cache_base.hh:285]: fix indentation

File: tests/03_generic_output/45_P0_data_DG_read.yaml
Comment (jbrezmorf) [reply tests/03_generic_output/45_P0_data_DG_read.yaml:1]: Cen we remove these `default_value`?

File: src/fields/field_fe.hh (hunk 4 above)
Comment (dflanderka) [reply src/fields/field_fe.hh]: comp_mesh_ removed

**Pull request merged**